floodlight/io/kinexon.py

```diff
@@ -0,0 +1,349 @@
+                # skip the header of the file
+                if line[parameter_links["time"]].isdecimal():
+                    # extract group id
+                    t.append(int(line[parameter_links["time"]]))
+                    if line[parameter_links["group_id"]] not in team_infos:
+                        team_infos.update({line[parameter_links["group_id"]]: {}})
+                    # create links
+                    for identifier in sensor_links:
+                        # extract identifier
+                        if (
+                            identifier
+                            not in team_infos[line[parameter_links["group_id"]]]
+                        ):
+                            team_infos[line[parameter_links["group_id"]]].update(
+                                {identifier: []}
+                            )
+                        # extract ids
+                        if (
+                            line[parameter_links[identifier]]
+                            not in team_infos[line[parameter_links["group_id"]]][
+                                identifier
+                            ]
+                        ):
+                            team_infos[line[parameter_links["group_id"]]][
+                                identifier
+                            ].append(line[parameter_links[identifier]])
+    # no group id in data
+    else:
+        print("Since no groups exist in data, artificial group '0' is created!")
+        with open(str(filepath_data), "r") as f:
+            while True:
+                line_string = f.readline()
+                # terminate if at end of file
+                if len(line_string) == 0:
+                    break
+                line = line_string.split(",")
+                # skip the header of the file
+                if line[parameter_links["time"]].isdecimal():
+                    t.append(int(line[parameter_links["time"]]))
+                    if "0" not in team_infos:
+                        team_infos.update({"0": {}})
+                    # create links
+                    for identifier in sensor_links:
+                        # extract identifier
+                        if identifier not in team_infos["0"]:
+                            team_infos["0"].update({identifier: []})
+                        # extract ids
+                        if (
+                            line[parameter_links[identifier]]
+                            not in team_infos["0"][identifier]
+                        ):
+                            team_infos["0"][identifier].append(
+                                line[parameter_links[identifier]]
+                            )
+
+    # sort dict
+    team_infos = dict(sorted(team_infos.items()))
+
+    # estimate framerate
+    timestamps = list(set(t))
+    timestamps.sort()
+    timestamps = np.array(timestamps)
+    minimum_time_step = np.min(np.diff(timestamps))
+    frame_rate = int(1000 / minimum_time_step)
```

Comment:
If the framerate, for whatever reason, ever turned out to be non-integer, this would automatically round it without notification. 

```
if not frame_rate.is_integer():
   warnings.warn(f"Non-integer frame rate: {frame_rate} detected, round to ....

Reply:
Whats the 1000? I assume you go from milliseconds to seconds (or something along this way). This might be one of the few cases where a magic number can be justified in the middle of the code, but that should be specified in a comment.

Reply:
Done